README.md
# JuMPChance

[![Build Status](https://travis-ci.org/mlubin/JuMPChance.jl.svg?branch=master)](https://travis-ci.org/mlubin/JuMPChance.jl)
[![Build Status](https://ci.appveyor.com/api/projects/status/github/mlubin/JuMPChance.jl?branch=master&svg=true)](https://ci.appveyor.com/project/mlubin/jumpchance-jl/branch/master)
[![Coverage Status](https://coveralls.io/repos/mlubin/JuMPChance.jl/badge.svg?branch=master&service=github)](https://coveralls.io/github/mlubin/JuMPChance.jl?branch=master)

JuMPChance (formerly CCJuMP) is an extension to [JuMP](https://github.com/JuliaOpt/JuMP.jl) for formulating and solving optimization problems with chance constraints (also known as probabilistic constraints). JuMPChance currently supports only a particular class of chance constraints involving affine combinations of jointly normal random variables, a classical formulation that's known to be efficiently solvable by using second-order conic programming (SOCP) (although JuMPChance also provides an outer-approximation algorithm which solves a sequence of linear problems).

JuMPChance supports an extension of the classical model to distributionally robust (or ambiguous) chance constraints where the parameters of the normal distributions are known to fall in a symmetric interval or more general uncertainty set.

See the [documentation](http://jumpchance.readthedocs.org/) for installation installation instructions, a quick start guide, and a more detailed discussion of the methods implemented.

Please cite JuMPChance using [![DOI](https://zenodo.org/badge/doi/10.5281/zenodo.13740.svg)](http://dx.doi.org/10.5281/zenodo.13740).


NEWS.md
JuMPChance release notes
========================

Version 0.2.2 (October 4, 2015)
-------------------------------

  * Fix deprecation warnings on Julia 0.4

Version 0.2.1 (September 3, 2015)
---------------------------------

  * Update to JuMP 0.10, new minimum required version.
  * ``getVar`` renamed to ``getVariance``
  * New ``satisfied_with_probability`` to check the probability violation at the current solution
  * Performance improvements
  * Special preprocessing for cases that can be reduced to linear constraints (thanks Kaarthik Sundar)
  * For two-sided chance constraints, choice between 2-approximation and 1.25 approximation

Version 0.2.0 (May 5, 2015)
---------------------------

  * Preliminary support for two-sided chance constraints.
  * ``solvechance`` method renamed to ``solve``
  * The meaning of ``with_probability`` **has changed**. Constraints must now hold with the given probability or greater, which is tractable for 1/2 or greater. A deprecation warning is in place when a small value is provided.
  * The ``@addConstraint`` macro now accepts chance constraints.
  * Following JuMP, the use of operators <= and >= *outside of macros* has been deprecated. Use ``@addConstraint`` instead of ``addConstraint``.

Version 0.1.1 (January 10, 2015)
--------------------------------

  * Clarify and tidy up support for integer variables.

Version 0.1.0 (January 4, 2015)
-------------------------------

  * Initial release


src/CCJuMP.jl
# extensions to JuMP for (robust) chance constraints


module CCJuMP
importall JuMP
import ECOS # For now, set ECOS as default since it's the only open-source conic solver availible

export CCModel,
    IndepNormal,
    affToStr,
    getMean,
    getVar,
    getStdev,
    solvecc,
    @defIndepNormal


# stores extension data inside JuMP Model
type CCData
    # pure chance constraints
    chanceconstr
    # robust chance constraints
    #robustchanceconstr
    
    numRVs::Int # number of independent normal r.v.'s
    RVmeans
    RVvars
    RVnames
end

function CCModel(;solver=ECOS.ECOSSolver())
    m = Model(solver=solver)
    m.ext[:ChanceConstr] = CCData(Any[],0,Any[],Any[],String[])
    return m
end

function getCCData(m::Model)
    if haskey(m.ext, :ChanceConstr)
        return m.ext[:ChanceConstr]::CCData
    else
        error("This functionality is only available for CCJuMP models")
    end
end


# pointer back to JuMP model and r.v. index
type IndepNormal
    m::Model
    idx::Int
end

function IndepNormal(m::Model, mean, var, name::String)
    ccdata = getCCData(m)
    ccdata.numRVs += 1
    push!(ccdata.RVmeans, mean)
    for v in var
        if v < 0
            error("Invalid value $v for variance of $name")
        end
    end
    push!(ccdata.RVvars, var)
    push!(ccdata.RVnames, name)
    return IndepNormal(m, ccdata.numRVs)
end

function getMean(v::IndepNormal)
    ccdata = getCCData(v.m)
    return ccdata.RVmeans[v.idx]
end

function getVar(v::IndepNormal)
    ccdata = getCCData(v.m)
    return ccdata.RVvars[v.idx]
end

getStdev(v::IndepNormal) = sqrt(getVar(v))

typealias CCAffExpr JuMP.GenericAffExpr{AffExpr,IndepNormal}

CCAffExpr() = CCAffExpr(IndepNormal[],AffExpr[],AffExpr())

Base.print(io::IO, a::CCAffExpr) = print(io, affToStr(a))
Base.show( io::IO, a::CCAffExpr) = print(io, affToStr(a))

# affine expression only involving r.v.'s
typealias RandomAffExpr JuMP.GenericAffExpr{Float64,IndepNormal}

RandomAffExpr() = RandomAffExpr(IndepNormal[],Float64[],0.0)

Base.print(io::IO, a::RandomAffExpr) = print(io, affToStr(a))
Base.show( io::IO, a::RandomAffExpr) = print(io, affToStr(a))

Base.promote_rule(::Type{CCAffExpr},::Type{RandomAffExpr}) = CCAffExpr
Base.promote_rule(::Type{RandomAffExpr},::Type{CCAffExpr}) = CCAffExpr

Base.convert(::Type{CCAffExpr},a::RandomAffExpr) = CCAffExpr(a.vars,[convert(AffExpr,c) for c in a.coeffs],convert(AffExpr,a.constant))



function affToStr(a::CCAffExpr)

    if length(a.vars) == 0
        return JuMP.aff_str(JuMP.REPLMode,a.constant)
    end
    
    m = a.vars[1].m
    ccdata = getCCData(m)

    v = IndexedVector(JuMP.AffExpr, 0)
    a = merge_duplicates(a, v, m)

    strs = ["($(JuMP.aff_str(JuMP.REPLMode,a.coeffs[i], show_constant=true)))*$(ccdata.RVnames[a.vars[i].idx])" for i in 1:length(a.vars)]
    return string(join(strs," + "), " + ", JuMP.aff_str(JuMP.REPLMode, a.constant, show_constant=true))
end


function affToStr(a::RandomAffExpr)

    if length(a.vars) == 0
        return string(a.constant)
    end
    
    m = a.vars[1].m
    ccdata = getCCData(m)

    v = IndexedVector(Float64, 0)
    a = merge_duplicates(a, v, m)

    strs = ["($(a.coeffs[i]))*$(ccdata.RVnames[a.vars[i].idx])" for i in 1:length(a.vars)]
    return string(join(strs," + "), " + ", string(a.constant))
end

type ChanceConstr
    ccexpr::CCAffExpr
    sense::Symbol # :(<=) or :(>=), right-hand side assumed to be zero
    with_probability::Float64 # with this probability *or less*
    uncertainty_budget_mean::Int # for now, with Bertsimas-Sim uncertainty
    uncertainty_budget_variance::Int # for now, with Bertsimas-Sim uncertainty
end

ChanceConstr(ccexpr::CCAffExpr,sense::Symbol) = ChanceConstr(ccexpr, sense, NaN, 0, 0)

function addConstraint(m::Model, constr::ChanceConstr; with_probability::Float64=NaN, uncertainty_budget_mean::Int=0, uncertainty_budget_variance::Int=0)
    if !(0 < with_probability < 1)
        error("Must specify with_probability between 0 and 1")
    end
    constr.with_probability = with_probability
    constr.uncertainty_budget_mean = uncertainty_budget_mean
    constr.uncertainty_budget_variance = uncertainty_budget_variance
    
    ccdata = getCCData(m)
    push!(ccdata.chanceconstr, constr)

end



function conToStr(c::ChanceConstr)
    s = "$(affToStr(c.ccexpr)) $(c.sense) 0"
    if isnan(c.with_probability)
        return s
    else
        return s*", with probability $(c.with_probability)"
    end
end

Base.print(io::IO, a::ChanceConstr) = print(io, conToStr(a))
Base.show( io::IO, a::ChanceConstr) = print(io, conToStr(a))

include("operators.jl")
include("macros.jl")
include("solve.jl")

end


src/macros.jl
using Base.Meta

import JuMP.@gendict

macro defIndepNormal(m, x, mean, var)
    m = esc(m)
    @assert isexpr(mean,:kw) && mean.args[1] == :mean
    @assert isexpr(var,:kw) && var.args[1] == :var
    mean = esc(mean.args[2])
    var = esc(var.args[2])

    if isa(x,Symbol)
        # easy case
        return quote
            $(esc(x)) = IndepNormal($m,$mean,$var,$(string(x)))
            nothing
        end
    else
        if !isexpr(x,:ref)
            error("Syntax error: Expected $var to be of form var[...]")
        end

        refcall, idxvars, idxsets, idxpairs, condition = JuMP.buildrefsets(x)
        varname = JuMP.getname(x)

        varstr = :(string($(string(varname)),"["))
        for idxvar in idxvars
            push!(varstr.args,:(string($(esc(idxvar)))))
            push!(varstr.args,",")
        end
        deleteat!(varstr.args,length(varstr.args))
        push!(varstr.args,"]")

        code = :( $(refcall) = IndepNormal($m, $mean, $var, $varstr ) )
        looped = JuMP.getloopedcode(x, code, condition, idxvars, idxsets, idxpairs, :IndepNormal)
        return quote
            $looped
            $(esc(varname))
        end
    end
end

# Extensions to make JuMP macros work with chance constraints

function JuMP.constructconstraint!(faff::CCAffExpr, sense::Symbol)
    if sense == :(<=) || sense == :≤
        return ChanceConstr(faff, :(<=))
    elseif sense == :(>=) || sense == :≥
        return ChanceConstr(faff, :(>=))
    elseif sense == :(==)
        error("Equality chance constraints not supported")
    end
    error("Unrecognized constraint type $sense")
end

JuMP.constructconstraint!(faff::CCAffExpr, lb::AffExpr, ub::AffExpr) = TwoSideChanceConstr(faff, lb, ub)
JuMP.constructconstraint!(faff::CCAffExpr, lb, ub) = TwoSideChanceConstr(faff, convert(AffExpr,lb), convert(AffExpr,ub))


src/operators.jl

import Base: +, -, /, *, <=, >=


# IndepNormal
(+)(lhs::IndepNormal,rhs::IndepNormal) = RandomAffExpr([lhs,rhs],[1.0,1.0],0.0)
(-)(lhs::IndepNormal,rhs::IndepNormal) = RandomAffExpr([lhs,rhs],[1.0,-1.0],0.0)
(/)(lhs::IndepNormal,rhs::Number) = RandomAffExpr([lhs],[1/rhs],0.0)
(-)(rhs::IndepNormal) = RandomAffExpr([rhs],[-1.0],0.0)


for op in (:+, :-, :*)
    @eval begin
        ($op)(lhs::Number, rhs::IndepNormal) = ($op)(lhs, RandomAffExpr([rhs],[1.0],0.0))
        ($op)(lhs::Variable, rhs::IndepNormal) = ($op)(convert(AffExpr,lhs),rhs)
        ($op)(lhs::Variable, rhs::RandomAffExpr) = ($op)(convert(AffExpr,lhs),rhs)
        ($op)(lhs::Variable, rhs::CCAffExpr) = ($op)(convert(AffExpr,lhs),rhs)
    end
    if op == :-
        @eval ($op)(lhs::CCAffExpr, rhs::Variable) = (+)(lhs,-rhs)
        @eval @compat ($op)(lhs::IndepNormal, rhs::Union{Variable,Number}) = (+)(lhs,-rhs)
    else
        @eval @compat ($op)(lhs::IndepNormal, rhs::Union{Variable,Number}) = ($op)(rhs,lhs)
        @eval ($op)(lhs::CCAffExpr, rhs::Variable) = ($op)(rhs,lhs)
    end
end


# AffExpr
# AffExpr--IndepNormal
(+)(lhs::AffExpr, rhs::IndepNormal) = CCAffExpr([rhs],[convert(AffExpr,1.0)],lhs)
(-)(lhs::AffExpr, rhs::IndepNormal) = CCAffExpr([rhs],[convert(AffExpr,-1.0)],lhs)
(*)(lhs::AffExpr, rhs::IndepNormal) = CCAffExpr([rhs],[lhs],AffExpr())
(+)(lhs::IndepNormal, rhs::AffExpr) = rhs+lhs
(-)(lhs::IndepNormal, rhs::AffExpr) = (+)(lhs,-rhs)
(*)(lhs::IndepNormal, rhs::AffExpr) = rhs*lhs

# AffExpr--CCAffExpr
(+)(lhs::AffExpr,rhs::CCAffExpr) = convert(CCAffExpr,lhs)+rhs
(+)(lhs::CCAffExpr,rhs::AffExpr) = rhs+lhs
(-)(lhs::AffExpr,rhs::CCAffExpr) = convert(CCAffExpr,lhs)-rhs
(-)(lhs::CCAffExpr,rhs::AffExpr) = lhs-convert(CCAffExpr,rhs)
Base.promote_rule(::Type{AffExpr},::Type{CCAffExpr}) = CCAffExpr
Base.convert(::Type{CCAffExpr},a::AffExpr) = CCAffExpr(IndepNormal[],AffExpr[],a)

# AffExpr--RandomAffExpr
(+)(lhs::AffExpr,rhs::RandomAffExpr) = lhs+convert(CCAffExpr,rhs)
(-)(lhs::AffExpr,rhs::RandomAffExpr) = lhs-convert(CCAffExpr,rhs)
(*)(lhs::AffExpr,rhs::RandomAffExpr) = CCAffExpr(rhs.vars, [lhs*c for c in rhs.coeffs], lhs*rhs.constant)

# CCAffExpr--RandomAffExpr
(+)(lhs::CCAffExpr,rhs::RandomAffExpr) = lhs+convert(CCAffExpr,rhs)
(-)(lhs::CCAffExpr,rhs::RandomAffExpr) = lhs-convert(CCAffExpr,rhs)

# CCAffExpr--CCAffExpr
# handled by GenericAffExpr fallback

# comparison operators
function (<=)(lhs::CCAffExpr, rhs::Number)
    Base.warn_once("The comparison operator <= has been deprecated for constructing chance constraints. Use the form @addConstraint instead.")
    ChanceConstr(lhs-rhs, :(<=))
end

function (>=)(lhs::CCAffExpr, rhs::Number)
    Base.warn_once("The comparison operator >= has been deprecated for constructing chance constraints. Use the form @addConstraint instead.")
    ChanceConstr(lhs-rhs, :(>=))
end
# == not valid

# RandomAffExpr
# RandomAffExpr--Variable
(+)(lhs::RandomAffExpr, rhs::Variable) = CCAffExpr(lhs.vars,[convert(AffExpr,c) for c in lhs.coeffs],rhs+lhs.constant)
(-)(lhs::RandomAffExpr, rhs::Variable) = CCAffExpr(lhs.vars,[convert(AffExpr,c) for c in lhs.coeffs],lhs.constant-rhs)
(*)(lhs::RandomAffExpr, rhs::Variable) = CCAffExpr(lhs.vars,[c*rhs for c in lhs.coeffs],rhs*lhs.constant)

# RandomAffExpr--AffExpr
(+)(lhs::RandomAffExpr,rhs::AffExpr) = rhs+lhs
(-)(lhs::RandomAffExpr,rhs::AffExpr) = (+)(-rhs,lhs)
(*)(lhs::RandomAffExpr,rhs::AffExpr) = rhs*lhs

# RandomAffExpr--CCAffExpr
(+)(lhs::RandomAffExpr,rhs::CCAffExpr) = convert(CCAffExpr,lhs)+rhs
(-)(lhs::RandomAffExpr,rhs::CCAffExpr) = convert(CCAffExpr,lhs)-rhs


src/distributions.jl
import Base.Math.@horner
# Copied from Distributions.jl because it takes too long to load


immutable Normal 
    μ::Float64
    σ::Float64

    function Normal(μ::Real, σ::Real)
    	σ > zero(σ) || error("std.dev. must be positive")
    	new(@compat(Float64(μ)), @compat(Float64(σ)))
    end

    Normal(μ::Real) = Normal(@compat(Float64(μ)), 1.0)
    Normal() = Normal(0.0, 1.0)
end

zval(d::Normal, x::Real) = (x - d.μ)/d.σ
xval(d::Normal, z::Real) = d.μ + d.σ * z

quantile(d::Normal, p::Real) = xval(d, Φinv(p))
cdf(d::Normal, x::Real) = Φ(zval(d,x))

Φ(z::Real) = 0.5*erfc(-z/√2)

# Rational approximations for the inverse cdf, from:
#   Wichura, M.J. (1988) Algorithm AS 241: The Percentage Points of the Normal Distribution
#   Journal of the Royal Statistical Society. Series C (Applied Statistics), Vol. 37, No. 3, pp. 477-484
for (fn,arg) in ((:Φinv,:p),(:logΦinv,:logp))
    @eval begin
        function $fn($arg::Real)
            if $(fn == :Φinv)
                q = p - 0.5
            else
                q = exp(logp) - 0.5
            end
            if abs(q) <= 0.425 
                r = 0.180625 - q*q
                return q * @horner(r,
                                   3.38713_28727_96366_6080e0, 
                                   1.33141_66789_17843_7745e2, 
                                   1.97159_09503_06551_4427e3, 
                                   1.37316_93765_50946_1125e4, 
                                   4.59219_53931_54987_1457e4, 
                                   6.72657_70927_00870_0853e4, 
                                   3.34305_75583_58812_8105e4, 
                                   2.50908_09287_30122_6727e3) /
                @horner(r,
                        1.0,
                        4.23133_30701_60091_1252e1, 
                        6.87187_00749_20579_0830e2, 
                        5.39419_60214_24751_1077e3, 
                        2.12137_94301_58659_5867e4, 
                        3.93078_95800_09271_0610e4, 
                        2.87290_85735_72194_2674e4, 
                        5.22649_52788_52854_5610e3)
            else
                if $(fn == :Φinv)
                    if p <= 0.0
                        return p == 0.0 ? -inf(Float64) : nan(Float64)
                    elseif p >= 1.0 
                        return p == 1.0 ? inf(Float64) : nan(Float64)
                    end
                    r = sqrt(q < 0 ? -log(p) : -log1p(-p))
                else
                    if logp == -Inf
                        return -inf(Float64)
                    elseif logp >= 0.0 
                        return logp == 0.0 ? inf(Float64) : nan(Float64)
                    end
                    r = sqrt(q < 0 ? -logp : -log1mexp(logp))
                end
                if r < 5.0
                    r -= 1.6
                    z = @horner(r,
                                1.42343_71107_49683_57734e0, 
                                4.63033_78461_56545_29590e0, 
                                5.76949_72214_60691_40550e0, 
                                3.64784_83247_63204_60504e0, 
                                1.27045_82524_52368_38258e0, 
                                2.41780_72517_74506_11770e-1, 
                                2.27238_44989_26918_45833e-2, 
                                7.74545_01427_83414_07640e-4) /
                    @horner(r,
                            1.0,
                            2.05319_16266_37758_82187e0, 
                            1.67638_48301_83803_84940e0, 
                            6.89767_33498_51000_04550e-1, 
                            1.48103_97642_74800_74590e-1, 
                            1.51986_66563_61645_71966e-2, 
                            5.47593_80849_95344_94600e-4, 
                            1.05075_00716_44416_84324e-9)
                else
                    r -= 5.0
                    z = @horner(r,
                                6.65790_46435_01103_77720e0, 
                                5.46378_49111_64114_36990e0, 
                                1.78482_65399_17291_33580e0, 
                                2.96560_57182_85048_91230e-1, 
                                2.65321_89526_57612_30930e-2, 
                                1.24266_09473_88078_43860e-3, 
                                2.71155_55687_43487_57815e-5, 
                                2.01033_43992_92288_13265e-7) /
                    @horner(r,
                            1.0,
                            5.99832_20655_58879_37690e-1, 
                            1.36929_88092_27358_05310e-1, 
                            1.48753_61290_85061_48525e-2, 
                            7.86869_13114_56132_59100e-4, 
                            1.84631_83175_10054_68180e-5, 
                            1.42151_17583_16445_88870e-7, 
                            2.04426_31033_89939_78564e-15)            
                end
                return copysign(z,q)
            end
        end
    end
end


src/JuMPChance.jl
# extensions to JuMP for (robust) chance constraints

isdefined(Base, :__precompile__) && __precompile__()

module JuMPChance
using JuMP
using Compat
import ECOS # For now, set ECOS as default since it's the only open-source conic solver availible

export ChanceModel,
    IndepNormal,
    affToStr,
    getMean,
    getVariance,
    getStdev,
    @defIndepNormal

# stores extension data inside JuMP Model
type CCData
    # pure chance constraints
    chanceconstr
    # robust chance constraints
    #robustchanceconstr
    # two-sided chance constraints
    twosidechanceconstr
    numRVs::Int # number of independent normal r.v.'s
    RVmeans
    RVvars
    RVnames
end

function ChanceModel(;solver=ECOS.ECOSSolver(verbose=false))
    m = Model(solver=solver)
    m.solvehook = solvehook
    m.ext[:ChanceConstr] = CCData(ChanceConstr[],TwoSideChanceConstr[],0,Any[],Any[],AbstractString[])
    return m
end

function getCCData(m::Model)
    if haskey(m.ext, :ChanceConstr)
        return m.ext[:ChanceConstr]::CCData
    else
        error("This functionality is only available for JuMPChance models")
    end
end


# pointer back to JuMP model and r.v. index
type IndepNormal <: JuMP.AbstractJuMPScalar
    m::Model
    idx::Int
end

function IndepNormal(m::Model, mean, var, name::AbstractString)
    ccdata = getCCData(m)
    ccdata.numRVs += 1
    push!(ccdata.RVmeans, mean)
    for v in var
        if v < 0
            error("Invalid value $v for variance of $name")
        end
    end
    push!(ccdata.RVvars, var)
    push!(ccdata.RVnames, name)
    return IndepNormal(m, ccdata.numRVs)
end

function getMean(v::IndepNormal)
    ccdata = getCCData(v.m)
    return ccdata.RVmeans[v.idx]
end

function getVariance(v::IndepNormal)
    ccdata = getCCData(v.m)
    return ccdata.RVvars[v.idx]
end

getStdev(v::IndepNormal) = sqrt(getVariance(v))

typealias CCAffExpr JuMP.GenericAffExpr{AffExpr,IndepNormal}

CCAffExpr() = CCAffExpr(IndepNormal[],AffExpr[],AffExpr())

Base.print(io::IO, a::CCAffExpr) = print(io, JuMP.affToStr(a))
Base.show( io::IO, a::CCAffExpr) = print(io, JuMP.affToStr(a))

# affine expression only involving r.v.'s
typealias RandomAffExpr JuMP.GenericAffExpr{Float64,IndepNormal}

RandomAffExpr() = RandomAffExpr(IndepNormal[],Float64[],0.0)

Base.print(io::IO, a::RandomAffExpr) = print(io, JuMP.affToStr(a))
Base.show( io::IO, a::RandomAffExpr) = print(io, JuMP.affToStr(a))

Base.promote_rule(::Type{CCAffExpr},::Type{RandomAffExpr}) = CCAffExpr
Base.promote_rule(::Type{RandomAffExpr},::Type{CCAffExpr}) = CCAffExpr

Base.convert(::Type{CCAffExpr},a::RandomAffExpr) = CCAffExpr(a.vars,[convert(AffExpr,c) for c in a.coeffs],convert(AffExpr,a.constant))



function JuMP.affToStr(a::CCAffExpr)

    if length(a.vars) == 0
        return JuMP.aff_str(JuMP.REPLMode,a.constant)
    end
    
    m = a.vars[1].m
    ccdata = getCCData(m)

    v = IndexedVector(JuMP.AffExpr, 0)
    counts = IndexedVector(Int, 0)
    a = merge_duplicates(a, v, counts, m)

    strs = ["($(JuMP.aff_str(JuMP.REPLMode,a.coeffs[i], true)))*$(ccdata.RVnames[a.vars[i].idx])" for i in 1:length(a.vars)]
    return string(join(strs," + "), " + ", JuMP.aff_str(JuMP.REPLMode, a.constant, true))
end


function JuMP.affToStr(a::RandomAffExpr)

    if length(a.vars) == 0
        return string(a.constant)
    end
    
    m = a.vars[1].m
    ccdata = getCCData(m)

    v = IndexedVector(Float64, 0)
    a = merge_duplicates(a, v, m)

    strs = ["($(a.coeffs[i]))*$(ccdata.RVnames[a.vars[i].idx])" for i in 1:length(a.vars)]
    return string(join(strs," + "), " + ", string(a.constant))
end

type ChanceConstr <: JuMP.JuMPConstraint
    ccexpr::CCAffExpr
    sense::Symbol # :(<=) or :(>=), right-hand side assumed to be zero
    with_probability::Float64 # with this probability *or greater*
    uncertainty_budget_mean::Int # for now, with Bertsimas-Sim uncertainty
    uncertainty_budget_variance::Int # for now, with Bertsimas-Sim uncertainty
end

ChanceConstr(ccexpr::CCAffExpr,sense::Symbol) = ChanceConstr(ccexpr, sense, NaN, 0, 0)

function JuMP.addConstraint(m::Model, constr::ChanceConstr; with_probability::Float64=NaN, uncertainty_budget_mean::Int=0, uncertainty_budget_variance::Int=0)
    if !(0 < with_probability < 1)
        error("Must specify with_probability between 0 and 1")
    end
    if with_probability < 0.5
        Base.warn_once("The meaning of with_probability has flipped! Constraints should be satisfied with given probability or greater. For backwards compatibility, this constraint will be interpreted in the old sense.")
        with_probability = 1-with_probability
        constr.sense = (constr.sense == :(>=)) ? :(<=) : :(>=)
    end
    constr.with_probability = with_probability
    constr.uncertainty_budget_mean = uncertainty_budget_mean
    constr.uncertainty_budget_variance = uncertainty_budget_variance
    
    ccdata = getCCData(m)
    push!(ccdata.chanceconstr, constr)
    return ConstraintRef{ChanceConstr}(m, length(ccdata.chanceconstr))

end

function satisfied_with_probability(r::ConstraintRef{ChanceConstr})
    m = r.m
    ccdata = getCCData(m)
    cc = ccdata.chanceconstr[r.idx]

    no_uncertains = all(x->isa(x,Real), ccdata.RVmeans) && all(x->isa(x,Real), ccdata.RVvars)
    if !no_uncertains
        error("satisfied_with_probability not implemented for distributionally robust constraints")
    end

    nterms = length(cc.ccexpr.vars)
    μ = getValue(cc.ccexpr.constant)
    σ² = 0.0
    for i in 1:nterms
        μ += getValue(cc.ccexpr.coeffs[i])*getMean(cc.ccexpr.vars[i])
        σ² += (getStdev(cc.ccexpr.vars[i])*getValue(cc.ccexpr.coeffs[i]))^2
    end
    σ = sqrt(σ²)
    d = Normal(μ,σ)
    if cc.sense == :(>=)
        return 1 - cdf(d, 0)
    else
        return cdf(d, 0)
    end
end

function JuMP.conToStr(c::ChanceConstr)
    s = "$(affToStr(c.ccexpr)) $(c.sense) 0"
    if isnan(c.with_probability)
        return s
    else
        return s*", with probability $(c.with_probability)"
    end
end

Base.print(io::IO, a::ChanceConstr) = print(io, JuMP.conToStr(a))
Base.show( io::IO, a::ChanceConstr) = print(io, JuMP.conToStr(a))

type TwoSideChanceConstr <: JuMP.JuMPConstraint
    ccexpr::CCAffExpr
    lb::AffExpr
    ub::AffExpr
    with_probability::Float64 # with this probability *or greater*
    approx::AbstractString # outer approximation model used for this constraint.
end

TwoSideChanceConstr(ccexpr::CCAffExpr,lb::AffExpr,ub::AffExpr) = TwoSideChanceConstr(ccexpr, lb, ub, NaN, "none")

function JuMP.addConstraint(m::Model, constr::TwoSideChanceConstr; with_probability::Float64=NaN, approx::AbstractString="")
    if !(0.5 < with_probability < 1)
        error("Must specify with_probability between 1/2 and 1")
    end
    constr.with_probability = with_probability
    if approx != "1.25" && approx != "2.0"
        error("approx paramater must be provided as \"1.25\" or \"2.0\". See documentation for further details")
    end
    constr.approx = approx

    ccdata = getCCData(m)
    push!(ccdata.twosidechanceconstr, constr)
    return ConstraintRef{TwoSideChanceConstr}(m, length(ccdata.twosidechanceconstr))

end

function JuMP.conToStr(c::TwoSideChanceConstr)
    s = "$(affToStr(c.lb)) <= $(affToStr(c.ccexpr)) <= $(affToStr(c.ub))"
    if isnan(c.with_probability)
        return s
    else
        return s*", with probability $(c.with_probability)"
    end
end

Base.print(io::IO, a::TwoSideChanceConstr) = print(io, JuMP.conToStr(a))
Base.show( io::IO, a::TwoSideChanceConstr) = print(io, JuMP.conToStr(a))

function satisfied_with_probability(r::ConstraintRef{TwoSideChanceConstr})
    m = r.m
    ccdata = getCCData(m)
    cc = ccdata.twosidechanceconstr[r.idx]

    nterms = length(cc.ccexpr.vars)
    μ = getValue(cc.ccexpr.constant)
    σ² = 0.0
    for i in 1:nterms
        μ += getValue(cc.ccexpr.coeffs[i])*getMean(cc.ccexpr.vars[i])
        σ² += (getStdev(cc.ccexpr.vars[i])*getValue(cc.ccexpr.coeffs[i]))^2
    end
    σ = sqrt(σ²)
    return cdf(Normal(μ,σ), getValue(cc.ub)) - cdf(Normal(μ,σ), getValue(cc.lb))

end

include("operators.jl")
include("macros.jl")
include("solve.jl")

end
